src/algebra/add_torsor.lean

```diff
@@ -0,0 +1,191 @@
+
+/-- An `add_torsor G P` gives a structure to the nonempty type `P`,
+acted on by an `add_group G` with a transitive and free action given
+by the `+ᵥ` operation and a corresponding subtraction given by the
+`-ᵥ` operation. In the case of a vector space, it is an affine
+space. -/
+class add_torsor (G : Type*) (P : Type*) [add_group G] extends add_action G P, has_vsub G P :=
+(nonempty : nonempty P)
+(vsub_vadd' : ∀ (p1 p2 : P), (p1 -ᵥ p2 : G) +ᵥ p2 = p1)
+(vadd_vsub' : ∀ (g : G) (p : P), g +ᵥ p -ᵥ p = g)
+end prio
+
+/-- An `add_group G` is a torsor for itself. -/
+instance add_group_is_add_torsor (G : Type*) [add_group G] :
+  add_torsor G G :=
+{ vadd := has_add.add,
+  vsub := has_sub.sub,
+  zero_vadd' := zero_add,
+  vadd_assoc' := λ a b c, (add_assoc a b c).symm,
+  nonempty := ⟨0⟩,
+  vsub_vadd' := sub_add_cancel,
+  vadd_vsub' := add_sub_cancel }
+
+namespace add_torsor
+
+section general
+
+variables (G : Type*) {P : Type*} [add_group G] [S : add_torsor G P]
+include S
+
+/-- Adding the zero group element to a point gives the same point. -/
+@[simp] lemma zero_vadd (p : P) : (0 : G) +ᵥ p = p :=
+zero_vadd' p
+
+/-- Adding two group elements to a point produces the same result as
+adding their sum. -/
+lemma vadd_assoc (g1 g2 : G) (p : P) : g1 +ᵥ (g2 +ᵥ p) = (g1 + g2) +ᵥ p :=
+vadd_assoc' g1 g2 p
+
+/-- Adding the result of subtracting from another point produces that
+point. -/
+@[simp] lemma vsub_vadd (p1 p2 : P) : (p1 -ᵥ p2 : G) +ᵥ p2 = p1 :=
+vsub_vadd' p1 p2
+
+/-- Adding a group element then subtracting the original point
+produces that group element. -/
+@[simp] lemma vadd_vsub (g : G) (p : P) : g +ᵥ p -ᵥ p = g :=
+vadd_vsub' g p
+
+/-- If the same point added to two group elements produces equal
+results, those group elements are equal. -/
+lemma vadd_cancel_right (g1 g2 : G) (p : P) (h : g1 +ᵥ p = g2 +ᵥ p) : g1 = g2 :=
+by rw [←vadd_vsub G g1, h, vadd_vsub]
+
+/-- Adding a group element to a point, then subtracting another point,
+produces the same result as subtracting the points then adding the
+group element. -/
+lemma vadd_vsub_assoc (g : G) (p1 p2 : P) : g +ᵥ p1 -ᵥ p2 = g + (p1 -ᵥ p2) :=
+begin
+  apply vadd_cancel_right G _ _ p2,
+  rw [vsub_vadd, ←vadd_assoc, vsub_vadd]
+end
+
+/-- Subtracting a point from itself produces 0. -/
+@[simp] lemma vsub_self (p : P) : p -ᵥ p = (0 : G) :=
+by rw [←zero_add (p -ᵥ p : G), ←vadd_vsub_assoc, vadd_vsub]
+
+/-- If subtracting two points produces 0, they are equal. -/
+lemma eq_of_vsub_eq_zero {p1 p2 : P} (h : p1 -ᵥ p2 = (0 : G)) : p1 = p2 :=
+by rw [←vsub_vadd G p1 p2, h, zero_vadd]
+
+/-- Subtracting two points produces 0 if and only if they are
+equal. -/
+lemma vsub_eq_zero_iff_eq (p1 p2 : P) : p1 -ᵥ p2 = (0 : G) ↔ p1 = p2 :=
+iff.intro (eq_of_vsub_eq_zero G) (λ h, h ▸ vsub_self G _)
+
+/-- Subtracting two points in the reverse order produces the negation
+of subtracting them. -/
+lemma vsub_rev_eq_neg_vsub (p1 p2 : P) : (p2 -ᵥ p1 : G) = -(p1 -ᵥ p2) :=
```

Comment:
I would swap the LHS and RHS of the equation, and make this `@[simp]`.

Reply:
Done.